### PR TITLE
Optimize Enumerable.count/1 implementation for File.Stream

### PR DESCRIPTION
**Pull request by michalmuskala**
Results of initial benchmarks using https://github.com/michalmuskala/file_count_optim
```
-##### With input Huge (1000000 lines) #####
Name                  ips        average  deviation         median
proposed             4.02      248.64 ms     ±3.62%      248.58 ms
recur_count          3.40      294.41 ms     ±4.37%      290.18 ms
read_line            1.22      817.86 ms     ±3.91%      806.50 ms
current              1.08      925.23 ms     ±5.24%      911.27 ms

Comparison:
proposed             4.02
recur_count          3.40 - 1.18x slower
read_line            1.22 - 3.29x slower
current              1.08 - 3.72x slower

-##### With input Medium (10000 lines) #####
Name                  ips        average  deviation         median
proposed           452.59        2.21 ms    ±11.83%        2.15 ms
recur_count        313.31        3.19 ms     ±6.88%        3.15 ms
read_line          109.97        9.09 ms     ±6.11%        8.94 ms
current            101.86        9.82 ms    ±14.58%        9.45 ms

Comparison:
proposed           452.59
recur_count        313.31 - 1.44x slower
read_line          109.97 - 4.12x slower
current            101.86 - 4.44x slower

-##### With input Small (100 lines) #####
Name                  ips        average  deviation         median
proposed          13.59 K       73.61 μs    ±29.75%       68.00 μs
recur_count       11.99 K       83.41 μs    ±27.13%       77.00 μs
read_line          6.49 K      154.07 μs    ±22.86%      142.00 μs
current            5.83 K      171.60 μs    ±41.22%      157.00 μs

Comparison:
proposed          13.59 K
recur_count       11.99 K - 1.13x slower
read_line          6.49 K - 2.09x slower
current            5.83 K - 2.33x slower
```

### CODE REVIEW

File: lib/elixir/lib/file/stream.ex [hunk 1]
```diff
@@ -36,7 +36,7 @@ defmodule File.Stream do
 
   defimpl Collectable do
     def into(%{path: path, modes: modes, raw: raw} = stream) do
-      modes = for mode <- modes, not mode in [:read], do: mode
+      modes = modes -- [:read]
```
Comment (michalmuskala) [inline lib/elixir/lib/file/stream.ex]: That's an unrelated change, but I think it simplifies the code and is small enough that it's ok to merge them together.

Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: its not quite equivalent as could be multiple `:read`

File: lib/elixir/lib/file/stream.ex [hunk 2]
```diff
@@ -88,12 +90,46 @@ defmodule File.Stream do
...
+      case File.open(path, modes, counter) do
+        {:ok, count} ->
+          {:ok, count}
+        {:error, reason} ->
+          raise File.Error, reason: reason, action: "stream", path: path
+      end
+    end
+
+    def count(%{path: path, line_or_bytes: bytes}) do
+      case File.stat(path) do
```
Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: do all files produce the correct size in stat?

Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: e.g. "/proc/uptime"

Comment (michalmuskala) [inline lib/elixir/lib/file/stream.ex]: It reports 0, but `File.read!("proc/uptime")` returns an empty string...

Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: What does `"/proc/uptime" |> File.stream!() |> Enum.to_list()` return? I think that one will work but can't remember how these edge cases play out.

Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: Yeah the stream versions works, it returns a list containing a single line.

Comment (michalmuskala) [inline lib/elixir/lib/file/stream.ex]: Yeah, that one works just fine.

What should we do here? I'm tempted to just return `{:error, __MODULE__}` in that case and keep the default implementation.

File: lib/elixir/lib/file/stream.ex (hunk 1 above)
Comment (michalmuskala) [inline lib/elixir/lib/file/stream.ex]: That's a good point. I'll revert the change

File: lib/elixir/lib/file/stream.ex (hunk 2 above)
Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: I think we should but if we lose the trivial O(1) count I would be indifferent to this getting in. Are we going to continue to add 30LOC optimisation to outlier functions like that make the code less clear?

Comment (fishcakez) [inline lib/elixir/lib/file/stream.ex]: @michalmuskala on you meant only if `size: 0`? I think that sounds good.

**Comment by michalmuskala** [comment]
@fishcakez updated

File: lib/elixir/lib/file/stream.ex [hunk 3]
```diff
@@ -88,12 +88,50 @@ defmodule File.Stream do
...
+        :eof ->
+          count
+        {:error, reason} ->
+          raise File.Error, reason: reason, action: "stream", path: path
+      end
+    end
+
+    defp count_lines(data, pattern), do: length(:binary.matches(data, pattern))
+
+    defp read_function(%{raw: true}),  do: &IO.binread(&1, @read_ahead_size)
```
Comment (lexmag) [inline lib/elixir/lib/file/stream.ex]: Unnecessary aligning of `do:`. :bowtie:

File: lib/elixir/lib/file/stream.ex [hunk 4]
```diff
@@ -88,12 +88,50 @@ defmodule File.Stream do
...
+          raise File.Error, reason: reason, action: "stream", path: path
+      end
     end
 
     def member?(_stream, _term) do
       {:error, __MODULE__}
     end
+
+    defp read_modes(modes),
+      do: for(mode <- modes, not mode in [:write, :append], do: mode)
```
Comment (lexmag) [inline lib/elixir/lib/file/stream.ex]: Perhaps it'd be better to use regular `do ... end` syntax here.

**Comment by michalmuskala** [comment]
@lexmag updated

**Comment by fishcakez** [comment]
Thank you :heart:

**Pull request merged**